src/pybind/mgr/volumes/fs/stats_util.py

```diff
@@ -57,3 +78,252 @@ def get_stats(src_path, dst_path, fs_handle, human=True):
+
+        if len(clones) < self.max_concurrent_clones:
+            total_clones = len(clones)
+        else:
+            total_clones = self.max_concurrent_clones
+
+        for clone in clones[:total_clones]:
+            with open_volume_lockless(self.volclient, clone.volname) as \
+                    fs_handle:
+                percent = get_percent_copied(clone.src_path, clone.dst_path,
+                                             fs_handle)
+                sum_percent += percent
+
+        avg_percent = round(sum_percent / total_clones, 3)
+        # progress module takes progress as a fraction between 0.0 to 1.0.
+        avg_progress_fraction = avg_percent / 100
+        msg = f'{total_clones} ongoing clones; Avg progress is {avg_percent}%'
+        self._update_progress_bar_event(ev_id=self.on_pev_id, ev_msg=msg,
+            ev_progress_fraction=avg_progress_fraction)
+
+        log.info('finished updating progress bar for ongoing clones')
+        return total_clones, sum_percent
+
+    def _update_progress_bars(self):
+        '''
+        Look for amount of progress made by all cloning operations and prints
+        progress bars, in "ceph -s" output, for average progress made
+        accordingly.
+
+        This method is supposed to be run only by instance of class RTimer
+        present in this class.
+        '''
+        clones = self._get_info_for_all_clones()
+
+        if clones:
+            total_clones, sum_percent = self._update_ongoing_progress_bar(clones)
+
+            # onpen bar (that is progress bar for clone jobs in ongoing and pending
+            # state) is printed when clones are in pending state. it is kept in
+            # printing until all clone jobs finish.
+            if len(clones) > self.max_concurrent_clones:
+                self._update_onpen_progress_bar(clones, total_clones, sum_percent)
+        else:
+            self._wait_and_finish()
+
+    def _finish_progress_events(self):
+        '''
+        Remove progress bars from "ceph status" output.
+        '''
+        log.info('removing progress bars from "ceph status" output')
+
+        assert self.on_pev_id is not None
+        assert self.onpen_pev_id is not None
+
+        self.volclient.mgr.remote('progress', 'complete', self.on_pev_id)
+        self.on_pev_id = None
+
+        self.volclient.mgr.remote('progress', 'complete', self.onpen_pev_id)
+        self.onpen_pev_id = None
+
+        log.info('finished removing progress bars from "ceph status" output')
+
+    def _found_clone_jobs_during_wait(self):
+        log.info('waiting for new clones before terminating this thread.')
+
+        interval = 1
+        wait_limit = 30
+        start_time = time()
+        cur_time = start_time
+        time_elapsed = 0
+
+        while time_elapsed < wait_limit:
+            if self._get_info_for_all_clones():
+                # found more clones, not finishing
+                return True
+
+            sleep(interval)
```

Comment:
This is not how waiting for more clones should be implemented. The timer thread should reschedule itself and then upon a certain number of iterations where no further clone activity is seen is when the timer task should shut down.

IMO, we can add this as a follow up since this whole change is getting closer to be tested and merged.

Reply:
Cool, I've removed code for this from current PR. Will create a separate PR for this.

Reply:
Marking this as resolved.